### ISSUE TITLE
7z2john.pl issue undefined value as a HASH reference

### ISSUE THREAD

**Issue by adamjk-dev** [openwall/john#4496]
### Error Message ###
──(kali㉿kali)-[~]
└─$ /usr/share/john/7z2john.pl /mnt/shared/archive.7z
DEBUG: number_coders: 2, codec_id: 
Can't use an undefined value as a HASH reference at /usr/share/john/7z2john.pl line 1384.
codec_id is:       

### Steps to reproduce ###
Latest kali Linux downloaded today and tested.
Was trying to get the hash out of an old 7-ZIP archive.

7-Zip details:
* When I test the 7zip archive it returns this:
Type = 7z
Physical Size = 56495367
Headers Size = 4771
Method = LZMA:24 7zAES
Solid = +
Blocks = 3

* Running file on the 7zip archive: 
7-zip archive data, version 0.3

### System configuration ###
──(kali㉿kali)-[~]
└─$ john --list=build-info                                                                                                               255 ⨯
Created directory: /home/kali/.john
Version: 1.9.0-jumbo-1
Build: linux-gnu 64-bit x86_64 AVX2 AC OMP
SIMD: AVX2, interleaving: MD4:3 MD5:3 SHA1:1 SHA256:1 SHA512:1
System-wide exec: /usr/lib/john
System-wide home: /usr/share/john
Private home: ~/.john
CPU tests: AVX2
CPU fallback binary: john-xop-omp
OMP fallback binary: john-avx2-non-omp
$JOHN is /usr/share/john/
Format interface version: 14
Max. number of reported tunable costs: 4
Rec file version: REC4
Charset file version: CHR3
CHARSET_MIN: 1 (0x01)
CHARSET_MAX: 255 (0xff)
CHARSET_LENGTH: 24
SALT_HASH_SIZE: 1048576
SINGLE_IDX_MAX: 32768
SINGLE_BUF_MAX: 4294967295
Effective limit: Max. KPC 32768
Max. Markov mode level: 400
Max. Markov mode password length: 30
gcc version: 9.2.1
GNU libc version: 2.28 (loaded: 2.31)
Crypto library: OpenSSL
OpenSSL library version: 01010103f      (loaded: 01010107f)
OpenSSL 1.1.1c  28 May 2019     (loaded: OpenSSL 1.1.1g  21 Apr 2020)
GMP library version: 6.1.2      (loaded: 6.2.0)
File locking: fcntl()
fseek(): fseek
ftell(): ftell
fopen(): fopen
memmem(): System's

┌──(kali㉿kali)-[~]
└─$ uname -a                                                                                                                               1 ⨯
Linux kali 5.9.0-kali1-amd64 #1 SMP Debian 5.9.1-1kali2 (2020-10-29) x86_64 GNU/Linux

**Comment by solardiz** [comment]
@adamjk-dev Please try to reproduce this issue using the latest script from this repo. We're not very interested in an issue that we might have already fixed (but still exists in a distro). Thanks!

**Comment by adamjk-dev** [comment]
Sure thing, looks like same result, but different line number:

```
──(kali㉿kali)-[~]
└─$ cd test
                                                                                                                                          
┌──(kali㉿kali)-[~/test]
└─$ pwd
/home/kali/test
                                                                                                                                          
┌──(kali㉿kali)-[~/test]
└─$ ll
total 0
                                                                                                                                          
┌──(kali㉿kali)-[~/test]
└─$ wget https://raw.githubusercontent.com/openwall/john/bleeding-jumbo/run/7z2john.pl
--2020-11-29 14:29:56--  https://raw.githubusercontent.com/openwall/john/bleeding-jumbo/run/7z2john.pl
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.184.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.184.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 86891 (85K) [text/plain]
Saving to: ‘7z2john.pl’

7z2john.pl                         100%[==============================================================>]  84.85K  --.-KB/s    in 0.03s   

2020-11-29 14:29:56 (2.92 MB/s) - ‘7z2john.pl’ saved [86891/86891]

                                                                                                                                          
┌──(kali㉿kali)-[~/test]
└─$ chmod +x 7z2john.pl 
                                                                                                                                          
┌──(kali㉿kali)-[~/test]
└─$ ./7z2john.pl /mnt/shared/archive.7z 
Can't use an undefined value as a HASH reference at ./7z2john.pl line 1433.
                                                                                                                                          
┌──(kali㉿kali)-[~/test]
└─$ md5sum 7z2john.pl                                                                                                               255 ⨯
10addb05d7e507d6ff030c5abd928967  7z2john.pl
                                                                                                                                          
┌──(kali㉿kali)-[~/test]
└─$ 

```

**Comment by solardiz** [comment]
Thanks. Are you able to create an archive that also triggers the issue but has non-sensitive content, so that you'd share it with us?

**Comment by adamjk-dev** [comment]
@solardiz Unknown at the moment. I think this is an old archive. I imagine it would have to be attempted to create a 7-Zip archive with version 0.3 etc. with the old encodings etc. I don't have a way to reproduce an archive at the moment.

**Comment by magnumripper** [comment]
All our samples show as "version 0.3" or "version 0.2" with file(1).

Perhaps `Solid = +` is the problem? Neither 7z2john or our formats has anything matching `grep -i solid`. Heads-up to @philsmd - our upstream is https://github.com/philsmd/7z2hashcat

**Comment by magnumripper** [comment]
I tried creating a solid archive (using `-ms=on`) and it was listed by file(1) as version 0.4. `7z2john.pl` had no problem with it, and our formats could crack it.

**Comment by magnumripper** [comment]
> * When I test the 7zip archive it returns this:
>   Type = 7z
>   Physical Size = 56495367
>   Headers Size = 4771
>   Method = LZMA:24 7zAES
>   Solid = +
>   Blocks = 3

There's nothing of the above we don't support.

**Comment by adamjk-dev** [comment]
Interesting, same result on that script as well...

```
──(kali㉿kali)-[~]
└─$ cd test
                                                                                                                                         
┌──(kali㉿kali)-[~/test]
└─$ ll
total 88
-rwxr-xr-x 1 kali kali 86891 Nov 29 14:29 7z2john.pl
                                                                                                                                         
┌──(kali㉿kali)-[~/test]
└─$ wget https://raw.githubusercontent.com/philsmd/7z2hashcat/master/7z2hashcat.pl
--2020-11-29 21:20:41--  https://raw.githubusercontent.com/philsmd/7z2hashcat/master/7z2hashcat.pl
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 199.232.96.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|199.232.96.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 86881 (85K) [text/plain]
Saving to: ‘7z2hashcat.pl’

7z2hashcat.pl                      100%[=============================================================>]  84.84K  --.-KB/s    in 0.04s   

2020-11-29 21:20:42 (1.86 MB/s) - ‘7z2hashcat.pl’ saved [86881/86881]

                                                                                                                                         
┌──(kali㉿kali)-[~/test]
└─$ chmod +x 7z2hashcat.pl 
                                                                                                                                         
┌──(kali㉿kali)-[~/test]
└─$ md5sum ./7z2hashcat.pl 
d769aa07d30172fde3afd7ee4be8107a  ./7z2hashcat.pl
                                                                                                                                         
┌──(kali㉿kali)-[~/test]
└─$ ./7z2hashcat.pl /mnt/shared/archive.7z 
Can't use an undefined value as a HASH reference at ./7z2hashcat.pl line 1433.
                                                                                                                                         
┌──(kali㉿kali)-[~/test]
└─$                       
```



At this point, do we consider debugging output statements or an enhanced logging level?

**Comment by adamjk-dev** [comment]
I ran the Perl debugger and set a breakpoint on line 1431 and dumped some variables (output was too long to include, but below is some of the last output):

```
┌──(kali㉿kali)-[~/test]
└─$ perl -d 7z2hashcat.pl archive.7z   

> b 1431

> c

> y

.... (SUPPRESSED OUTPUT)


$compression_attributes = '5d00000001'
$crc = 1521139124
$crc_len = 1158738
$current_index = 2
$current_seek_position = 1174628
$data = undef
$data_len = 55316000
$data_offset = 0
$data_offset_tmp = 1174458
$digest = HASH(0x5639bfdf6c90)
   'crc' => 2988940001
   'defined' => 1
$digests = ARRAY(0x5639bfdba070)
     empty array
$digests_index = 2
$file_idx = 1
$file_path = 'archive.7z'

$has_encrypted_header = 0
$hash_buf = ''
$is_truncated = 0
$iv_buf = "1*=\c?�e\c?a\c@\c@\c@\c@\c@\c@\c@\c@"
$iv_len = 8
$length_compressed = 138
$length_first_file = 151
$memory_buffer_read_offset = 20938
$next_file_idx = 1
$number_coders = 2
$number_cycles_power = 19
$number_file_indices = 267
$number_folders = 3
$number_pack_info = 3
$number_streams = 3


'number_digests' => 267
   'unpack_sizes' => ARRAY(0x5639bfc918e0)
      0  1158738
      1  151
      2  95006
      3  191641
      4  101229
      5  233481
      6  218021
      7  190340
      8  203525
      9  87759
      10  163937
      11  128421
      12  105201
      13  178920
      14  189035
      15  95505
      16  103139
      17  132583
      18  86868
      19  188112
      20  187254
      21  230414
      22  163207
      23  93844
      24  146089
      25  236166
      26  203573
      27  194495
      28  201750
      29  136480
      30  144606
      31  184880
      32  229608
      33  190008
      34  198692
      35  229930
      36  203562
      37  207724
      38  144057
      39  84260
      40  237436
      41  246246
      42  91236
      43  180992
      44  212913
      45  90894
      46  174100
      47  168471
      48  94899
      49  283402
      50  258259
      51  94466
      52  234459
      53  160261
      54  214646
      55  88970
      56  173750
      57  224021
      58  78446
      59  214061
      60  92909
      61  136074
      62  201386
      63  145648
      64  142503
      65  221691
      66  139045
      67  208189
      68  275389
      69  201380
      70  197027
      71  168096
      72  135898
      73  223254
      74  251288
      75  159328
      76  180399
      77  237530
      78  77917
      79  169235
      80  178826
      81  238024
      82  136717
      83  224332
      84  272375
      85  165902
      86  81478
      87  89079
      88  216239
      89  148794
      90  155707
      91  215410
      92  92545
      93  215433
      94  101552
      95  82945
      96  269315
      97  248330
      98  93055
      99  240967
      100  92520
      101  228967
      102  181402
      103  175229
      104  213347
      105  217504
      106  89714
      107  125840
      108  257340
      109  271974
      110  162236
      111  173465
      112  126070
      113  72120
      114  186654
      115  79239
      116  246024
      117  200443
      118  241116
      119  149978
      120  190320
      121  80575
      122  263774
      123  187461
      124  135607
      125  236526
      126  181272
      127  194219
      128  148148
      129  93031
      130  284920
      131  85864
      132  144972
      133  167161
      134  226094
      135  91570
      136  245871
      137  174018
      138  237725
      139  215378
      140  189983
      141  432284
      142  81915
      143  187558
      144  146842
      145  149328
      146  154250
      147  248543
      148  170809
      149  247798
      150  169727
      151  261082
      152  158480
      153  164420
      154  220224
      155  200601
      156  174400
      157  184831
      158  154543
      159  136719
      160  217268
      161  265064
      162  89843
      163  230321
      164  211825
      165  142770
      166  189440
      167  95964
      168  162422
      169  234360
      170  167161
      171  82978
      172  211207
      173  119685
      174  154408
      175  88544
      176  251157
      177  162998
      178  136574
      179  91598
      180  158429
      181  172290
      182  137524
      183  173754
      184  179713
      185  204906
      186  88344
      187  86390
      188  156475
      189  189754
      190  134731
      191  162343
      192  173349
      193  148858
      194  97456
      195  145928
      196  191041
      197  222064
      198  194737
      199  185091
      200  89449
      201  187652
      202  260884
      203  166913
      204  208597
      205  254296
      206  26368
      207  136020
      208  50426
      209  101697
      210  54232
      211  101191
      212  110146
      213  155923
      214  130869
      215  131436
      216  119567
      217  124851
      218  139379
      219  135688
      220  120551
      221  125365
      222  122449
      223  117728
      224  148006
      225  106338
      226  116145
      227  58542
      228  133950
      229  141353
      230  106591
      231  45138
      232  116317
      233  129222
      234  103212
      235  1478456
      236  107608
      237  483899
      238  607012
      239  538545
      240  536607
      241  574694
      242  545417
      243  531147
      244  510743
      245  546044
      246  583773
      247  502458
      248  501804
      249  503991
      250  504791
      251  470940
      252  397777
      253  492031
      254  504141
      255  502631
      256  489110
      257  473466
      258  583658
      259  569150
      260  514743
      261  569099
      262  582952
      263  585459
      264  549797
      265  451727
      266  426829
   'unpack_stream_numbers' => ARRAY(0x5639bfc91868)
      0  1
      1  1
      2  265
$type_of_compression = 1
$type_of_data = 1
$type_of_preprocessor = 0
$unpack_info = HASH(0x5639bfdba8c8)
   'coder_unpack_sizes' => ARRAY(0x5639bfdb9fb0)
      0  0
      1  1
      2  2
   'datastream_indices' => ARRAY(0x5639bfdb9fc8)
        empty array
   'digests' => ARRAY(0x5639bfdba070)
        empty array
   'folders' => ARRAY(0x5639bfdba058)
      0  HASH(0x5639bfdbabe0)
         'bindpairs' => ARRAY(0x5639bfdb9e90)
              empty array
         'coders' => ARRAY(0x5639bfdb9ef0)
            0  HASH(0x5639bfdba880)
               'attributes' => "]\c@\c@\cX\c@"
               'codec_id' => "\cC\cA\cA"
               'number_input_streams' => 1
               'number_output_streams' => 1
         'index_main_stream' => 0
         'number_coders' => 1
         'sum_input_streams' => 1
         'sum_output_streams' => 1
         'sum_packed_streams' => 1
      1  HASH(0x5639bfdba3a0)
         'bindpairs' => ARRAY(0x5639bfdbac28)
              empty array
         'coders' => ARRAY(0x5639bfdba970)
            0  HASH(0x5639bfdbaa60)
               'attributes' => "]\c@\c@\cA\c@"
               'codec_id' => "\cC\cA\cA"
               'number_input_streams' => 1
               'number_output_streams' => 1
         'index_main_stream' => 0
         'number_coders' => 1
         'sum_input_streams' => 1
         'sum_output_streams' => 1
         'sum_packed_streams' => 1
      2  HASH(0x5639bfdba550)
         'bindpairs' => ARRAY(0x5639bfdba748)
            0  ARRAY(0x5639bfc7a130)
               0  1
               1  0
         'coders' => ARRAY(0x5639bfda7748)
            0  HASH(0x5639bfdba430)
               'attributes' => "S\cG1*=\c?�e\c?a"
               'codec_id' => "\cF�\cG\cA"
               'number_input_streams' => 1
               'number_output_streams' => 1
            1  HASH(0x5639bfdba4f0)
               'attributes' => "]\c@\c@\c@\cA"
               'codec_id' => "\cC\cA\cA"
               'number_input_streams' => 1
               'number_output_streams' => 1
         'index_main_stream' => 1
         'number_coders' => 2
         'sum_input_streams' => 2
         'sum_output_streams' => 2
         'sum_packed_streams' => 1
   'main_unpack_size_index' => ARRAY(0x5639bfdadb70)
      0  0
      1  0
      2  1
   'number_folders' => 3
   'unpack_sizes' => ARRAY(0x5639bfdba0a0)
      0  1158738
      1  151
      2  55315997
      3  55872240
$unpack_size = 55315997
%SEVEN_ZIP_COMPRESSOR_NAMES = (
   1 => 'LZMA1'
   112 => 'SPARC'
   16 => 'BCJ'
   2 => 'LZMA2'
   3 => 'PPMD'
   32 => 'BCJ2'
   48 => 'PPC'
   6 => 'BZIP2'
   64 => 'IA64'
   7 => 'DEFLATE'
   80 => 'ARM'
   96 => 'ARMT'
)
@PASSWORD_RECOVERY_TOOL_SUPPORTED_DECOMPRESSORS = (
   0  1
   1  2
)
@PASSWORD_RECOVERY_TOOL_SUPPORTED_PREPROCESSORS = (
   0  0
)
  DB<2> ^Cmain::extract_hash_from_archive(7z2hashcat.pl:1433):
1433:                 my $codec_id = @$coders[$coders_idx]->{'codec_id'};

```

**Comment by magnumripper** [comment]
So at that time, is `$coders_idx` 0 or 1?

**Comment by magnumripper** [comment]
Given that debug output I can't see how it ends up with this error. The loop at line 1431 should go from 0 through 1, as we saw `number_coders` at 2. We've also seen the `coders` array having everything defined that should be needed.

In short, I don't get it!

**Comment by magnumripper** [comment]
> In short, I don't get it!

I would try this: Edit line 113 and set `$ANALYZE_ALL_STREAMS_TO_FIND_SHORTEST_DATA_BUF` to 0, then try again (using 7z2john,pl, not 7z2hashcat.pl). Does it happen to end up better?

**Comment by magnumripper** [comment]
Also, try `7z -slt l /mnt/shared/archive.7z` and post the full output (unless it's very long). You can redact some data such as file names if needed.

**Comment by adamjk-dev** [comment]
So, setting `$ANALYZE_ALL_STREAMS_TO_FIND_SHORTEST_DATA_BUF = 0` allowed it to run.

Hashcat doesn't like it as is, but I will try running it in JTR I presume.
```
┌──(kali㉿kali)-[~/test]
└─$ hashcat -m 11600 hash.txt /usr/share/wordlists/rockyou.txt
hashcat (v6.1.1) starting...

OpenCL API (OpenCL 1.2 pocl 1.5, None+Asserts, LLVM 9.0.1, RELOC, SLEEF, DISTRO, POCL_DEBUG) - Platform #1 [The pocl project]
=============================================================================================================================
* Device #1: pthread-AMD Ryzen 5 1600 Six-Core Processor, 2889/2953 MB (1024 MB allocatable), 4MCU

Minimum password length supported by kernel: 0
Maximum password length supported by kernel: 256

Counted lines in hash.txt...
Oversized line detected! Truncated 1712638 bytes

Oversized line detected! Truncated 1712638 bytes
Hashfile 'hash.txt' on line 1 ($7z$1$...b02412bf8111a63f7b4ea42e81308f8f): Token length exception
No hashes loaded.

Started: Mon Nov 30 07:32:30 2020
Stopped: Mon Nov 30 07:32:31 2020

```

JTR aborted it seems:
```
┌──(kali㉿kali)-[~/test]
└─$ john --wordlist=/usr/share/wordlists/rockyou.txt hash.txt 
Using default input encoding: UTF-8
Loaded 1 password hash (7z, 7-Zip [SHA256 256/256 AVX2 8x AES])
Cost 1 (iteration count) is 524288 for all loaded hashes
Cost 2 (padding size) is 0 for all loaded hashes
Cost 3 (compression type) is 1 for all loaded hashes
Will run 4 OpenMP threads
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:52 0.01% (ETA: 2020-12-05 01:15) 0g/s 42.34p/s 42.34c/s 42.34C/s cedric..sprite
Session aborted
```

The size of the hash file is 2433542 bytes.

**Comment by solardiz** [comment]
JtR only prints "Session aborted" when it thinks you asked it to stop by pressing `q` or Ctrl-C or by sending it a signal. This must be unrelated to file contents.

**Comment by adamjk-dev** [comment]
Makes sense, I am running it again now. Not sure why hashcat hates it, but am trying JtR. Either way, you have both been super helpful and setting that variable did seem to let it through to create some hashes to work with it looks like. Have to try working with these.

**Comment by solardiz** [comment]
I am also puzzled by the discrepancy between "AMD Ryzen 5 1600 Six-Core Processor" and "4 threads". Are you perhaps running this in a VM configured to only have 4 vCPUs?

For your actual use, I recommend that you build the latest code from this repo, not use Kali's older version. @magnumripper made many improvements to the 7z format since the 1.9.0-jumbo-1 release, some of those include support for additional archive types - might be relevant for yours or not.

Thank you for your help debugging the Perl script issue, even though we're yet to fully figure it out and fix it.

**Comment by adamjk-dev** [comment]
Correct, a VM as stated with 4 vCPU. Indeed, I may need to port over to main machine etc. and let it crank there.

Good idea, I need to look into that.

Indeed, yeah I am not sure, I can try to go back and debug more too and see which loop iteration it failed etc. I am not particularly good in the Perl debugger, but can play around and see what was an invalid reference I would hope.

**Comment by magnumripper** [comment]
The `Cost 2 (padding size) is 0` isn't optimal - a padding size of at least 1 may boost your speed significantly. With some luck it would have been larger with `$ANALYZE_ALL_STREAMS_TO_FIND_SHORTEST_DATA_BUF` set to 1 as default.

If you feel like hacking about, you could try to let it find shorter streams but somehow test/skip whatever stream is failing.

BTW, a `7z -slt l /mnt/shared/archive.7z` would still be interesting for us. If output is large you can skipping repeating patterns. Perhaps just a `7z -slt l /mnt/shared/archive.7z | grep ^Method | uniq`.

**Comment by magnumripper** [comment]
Perhaps try this diff and see if it can get past the culprit
```diff
diff --git a/run/7z2john.pl b/run/7z2john.pl
index dfbb93cf4..0ea247beb 100755
--- a/run/7z2john.pl
+++ b/run/7z2john.pl
@@ -1430,7 +1430,17 @@ sub extract_hash_from_archive
 
             for (my $coders_idx = 0; $coders_idx < $number_coders; $coders_idx++)
             {
-              my $codec_id = @$coders[$coders_idx]->{'codec_id'};
+              my $codec_id;
+
+              eval
+              {
+                $codec_id = @$coders[$coders_idx]->{'codec_id'};
+              };
+              if ($@)
+              {
+                print STDERR "Stream $coders_idx no codec, recovering\n";
+                next;
+              }
 
               if ($codec_id eq $SEVEN_ZIP_AES)
               {
```

**Comment by adamjk-dev** [comment]
Here is the output from the uniq on Method:

```
┌──(kali㉿kali)-[~/test]
└─$ 7z -slt l archive.7z | grep ^Method | uniq 
Method = LZMA:24 7zAES
Method = LZMA:1536k
Method = LZMA:16
Method = LZMA:24 7zAES:19
Method = 
```

There are a lot of results (files within), so the output of the other call is quite large indeed.

**Comment by adamjk-dev** [comment]
For trying your patch code (and setting back the original variable -- ANALYZE_ALL_STREAMS_TO_FIND_SHORTEST_DATA_BUF -- to 1):
```
┌──(kali㉿kali)-[~/test]
└─$ ./7z2johnmod.pl archive.7z
Stream 1 no codec, recovering
ATTENTION: the hashes might contain sensitive encrypted data. Be careful when sharing or posting these hashes
archive.7z:$7z$1$19$0$$8$312a3d7fc8657f610000000000000000<REDACTED>.....

──(kali㉿kali)-[~/test]
└─$ ./7z2johnmod.pl archive.7z > new_hash.txt
Stream 1 no codec, recovering
ATTENTION: the hashes might contain sensitive encrypted data. Be careful when sharing or posting these hashes

┌──(kali㉿kali)-[~/test]
└─$ ls -lh new_hash.txt 
-rw-r--r-- 1 kali kali 477 Nov 30 12:52 new_hash.txt

```

So, it did output a much smaller hash this time if that is expected.

**Comment by magnumripper** [comment]
The `Method = ` last line looks like the culprit...

Perhaps this fix is enough
```diff
diff --git a/run/7z2john.pl b/run/7z2john.pl
index dfbb93cf4..786663ee7 100755
--- a/run/7z2john.pl
+++ b/run/7z2john.pl
@@ -1430,6 +1430,8 @@ sub extract_hash_from_archive
 
             for (my $coders_idx = 0; $coders_idx < $number_coders; $coders_idx++)
             {
+              next unless defined @$coders[$coders_idx];
+
               my $codec_id = @$coders[$coders_idx]->{'codec_id'};
 
               if ($codec_id eq $SEVEN_ZIP_AES)
```

**Comment by magnumripper** [comment]
> So, it did output a much smaller hash this time if that is expected.

Great, let's hope it has at least a little padding too :-)

**Comment by adamjk-dev** [comment]
Yep, same result for that patch code obviously. Any comment on the smaller hash result?

When we set that variable to 1 my hash size was 2433542 bytes vs. this one which is 477 bytes long.

Sorry, side question, for running against JtR, we don't need to remove the first couple of characters from the resultant hash for the file name, correct? Just run the result of this script against JtR as is?

**Comment by magnumripper** [comment]
> Yep, same result for that patch code obviously. Any comment on the smaller hash result?

Smaller is faster to attack, and if we also got a padding size > 0 it will be even faster!

> Sorry, side question, for running against JtR, we don't need to remove the first couple of characters from the resultant hash for the file name, correct? Just run the result of this script against JtR as is?

Yessir, it will even try to make some guesses from the file name if you keep it and run single (or batch) mode.

**Comment by adamjk-dev** [comment]
Okay, thanks again for the input and special attention, always a great thing to see. I am going to see if JtR can crack this smaller hash I just go out of it (the 477 bytes).

If you need me to do any more debugging etc. I can to help with any potential code changes you might make. Or, perhaps you are happy with the extra check we just tested and you suggested on the patch commands.

**Comment by magnumripper** [comment]
@adamjk-dev did you see better speed now? I think you should. Also, if you ever crack it please report here. While I'm pretty confident the fix is correct it's always nice to have it confirmed.

**Comment by adamjk-dev** [comment]
@magnumripper Thanks for checking in, still running it in that silly VM, so speeds aren't great (also running hashcat and JtR against it in 2 tabs on 4 vCPUs). So, it is still cranking haha, around 54% on each. I will let you know for sure. I may need to customize a ruleset or not run a wordlist etc., we shall see. I messed around installing it on Cygwin on my machine but hit some issues (likely on my end, I don't use this machine much). I need to try elsewhere too on a real machine. I am just being lame about it right now to be honest.

**Comment by solardiz** [comment]
@adamjk-dev This is kind of off-topic, but FYI both OpenMP and OpenCL assume an otherwise-idle system, and start wasting time on busy-waits when there's other load, even very slight. So running two crackers at once each thinking it has 4 vCPUs all to itself isn't a good idea, and it's hard to tell how much CPU is actually being used for computation rather than for busy-waits on thread synchronization then. Obviously, the busy-waits are reported by the system as CPU usage, just like real computation would be, just without making as much real progress. For JtR, a workaround is to fully switch to using `--fork` instead of OpenMP, in which case it uses out-of-sync child processes instead, without wasting time. Or indeed to keep the system otherwise idle. Or to keep enough logical CPUs otherwise-idle for the specified `OMP_NUM_THREADS`.

**Comment by adamjk-dev** [comment]
Appreciate it. My first runs on the VM didn't crack it via JtR and the rockyou wordlist, nor did hashcat on the same wordlist. I am running JtR on a full 8 cores on another machine by itself, no wordlist, and am trying some custom word lists etc. Just wanted to update, continuing to try.

**Comment by magnumripper** [comment]
I just now opened a PR with our fix upstream, closing this

https://github.com/philsmd/7z2hashcat/pull/32
